engine/src/OpenGL.cpp

```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+        CHECKED_GL_CALL(glEnableVertexAttribArray, 0);
+        CHECKED_GL_CALL(glVertexAttribPointer, 0, 2, GL_FLOAT, GL_FALSE, 4 * sizeof(float), (void *) 0);
+        CHECKED_GL_CALL(glEnableVertexAttribArray, 1);
+        CHECKED_GL_CALL(glVertexAttribPointer, 1, 2, GL_FLOAT, GL_FALSE, 4 * sizeof(float),
+                        (void *) (2 * sizeof(float)));
+
+        return vao;
+    }
+
+    // stencil to texture functions
+    void OpenGL::enable_stencil_testing() {
+        CHECKED_GL_CALL(glEnable, GL_STENCIL_TEST);
+    }
+    void OpenGL::disable_stencil_writing() {
+        CHECKED_GL_CALL(glStencilMask, 0x00);
+    }
+    void OpenGL::stencil_func(std::string func, int ref, unsigned int mask) {
+        GLenum func_;
+
+        if( "never" == func) func_ = GL_NEVER;
+        else if("less" == func) func_ = GL_LESS;
+        else if("equal" == func) func_ = GL_EQUAL;
+        else if("notequal" == func) func_ = GL_NOTEQUAL;
+        else if("greater" == func) func_ = GL_GREATER;
+        else if ("always" == func) func_ = GL_ALWAYS;
+        else func_ = GL_ALWAYS;
+
+        CHECKED_GL_CALL(glStencilFunc, func_, ref, mask);
+    }
+    void OpenGL::stencil_op(std::string sfail, std::string dfail, std::string dpass) {
+        GLenum sfail_;
+        if("keep" == sfail) sfail_ = GL_KEEP;
+        else if("zero" == sfail) sfail_ = GL_ZERO;
+        else if("replace" == sfail) sfail_ = GL_REPLACE;
+        else if("incr" == sfail) sfail_ = GL_INCR;
+        else if("invert" == sfail) sfail_ = GL_INVERT;
+        else if ("decr" == sfail) sfail_ = GL_DECR;
+
+        GLenum dfail_;
+
+        if("keep" == dfail) dfail_ = GL_KEEP;
+        else if("zero" == dfail) dfail_ = GL_ZERO;
+        else if("replace" == dfail) dfail_ = GL_REPLACE;
+        else if("incr" == dfail) dfail_ = GL_INCR;
+        else if("invert" == dfail) dfail_ = GL_INVERT;
+        else if ("decr" == dfail) dfail_ = GL_DECR;
+
+        GLenum dpass_;
+
+        if("keep" == dpass) dpass_ = GL_KEEP;
+        else if("zero" == dpass) dpass_ = GL_ZERO;
+        else if("replace" == dpass) dpass_ = GL_REPLACE;
+        else if("incr" == dpass) dpass_ = GL_INCR;
+        else if("invert" == dpass) dpass_ = GL_INVERT;
+        else if ("decr" == dpass) dpass_ = GL_DECR;
+
+        CHECKED_GL_CALL(glStencilOp, sfail_, dfail_, dpass_);
+    }
+    void OpenGL::stencil_mask(int mask) {
+        if (mask == 0)
+            CHECKED_GL_CALL(glStencilMask, 0x00);
+        else if (mask == 1)
+            CHECKED_GL_CALL(glStencilMask, 0xFF);
+    }
+
+    unsigned int OpenGL::copy_stencil_to_texture() {
+        auto platform = engine::core::Controller::get<engine::platform::PlatformController>();
+        int width     = platform->window()->width();
+        int height    = platform->window()->height();
+        unsigned int stencil_texture;
+        CHECKED_GL_CALL(glGenTextures, 1, &stencil_texture);
+        CHECKED_GL_CALL(glBindTexture, GL_TEXTURE_2D, stencil_texture);
```

Comment:
Čini mi se da ovaj pristup možda nije dovoljno fleksibilan ni efikasan, jer će u svakom frejmu biti kreiran novi OpenGL objekat.

Možda bi bilo bolje omogućiti prosleđivanje src i dst kopiranja kao argumente ove funkcije, slično kao što to radi memcpy. To bi moglo poboljšati efikasnost i omogućiti veću kontrolu nad procesom kopiranja